### DISCUSSION TITLE
Passing Parameters to GH Actions (From a Variable)

### DISCUSSION THREAD

**Discussion by eiymba** [community/community#26858] [Actions]
<p>Hi!</p>
<p>I’m looking for a way to pass parameters to GitHub Actions without hard coding it into the yaml file. This is because I’d like to reuse the yaml across projects. Is this possible?</p>
<h2>For Context</h2>
<p>my yaml file builds a docker image and publishes it to the GitHub Container Registry and it works fine. However, currently I’m using the repository name for the name of the docker image.</p>
<p>I’d like to arbitrarily set this value <strong>without writing it in the yaml file</strong>.</p>
<p>With Travis CI, you have the ability to set environment variables per repository in the dashboard, then use that variable.</p>
<p>Is this possible with GitHub Actions? I have an alternative solution regardless; I’m just wondering if this is possible before trying something else.</p>
<p>Help is appreciated, thanks!</p>

**Comment by airtower-luna** [comment]
<p>In principle you could use <a href="https://docs.github.com/en/actions/reference/encrypted-secrets">secrets</a> to store your variables.</p>
<p>However, as long as you need only the repository name you can get that from the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context"><code>github</code> context</a>, specifically <code>github.repository</code> (you might want to remove the owner prefix, depending on the desired output).</p>

**Comment by eiymba** [comment]
<p>Hi! Thanks for the reply!</p>
<p>Yeah, I tried this previous but unfortunately it didn’t work unfortunately for some reason.</p>
<p>It does not appear to let you use secrets arbitrarily like that.</p>
<p>the docker image format is <a href="http://ghcr.io/:organisation/:image-name::version" rel="noopener nofollow ugc">ghcr.io/:organisation/:image-name::version</a>, where <code>:</code> denotes a wildcard (except for version tag).</p>
<p>e.g.:</p>
<pre><code class="lang-auto">ghcr.io/${{ steps.get_owner.outputs.VALUE }}/${{ steps.release_name.outputs.VALUE }}:${{ steps.get_version.outputs.VALUE }}
</code></pre>
<p>This will comfortably release a docker image who’s name matches the repository.</p>
<p>the <em>release_name</em> step is:</p>
<pre><code class="lang-auto"> echo ::set-output name=VALUE::$(if [ "${{ secrets.RELEASE_NAME }}" ]; then echo ${{ secrets.RELASE_NAME }}; else echo "$GITHUB_REPOSITORY" | awk -F / '{print $2}'; fi)
</code></pre>
<p>Which executes fine, but cannot be used in the build step because it won’t output secrets like that, I think.</p>
<p>Give us a shout if you come up with anything else. Thanks again!</p>

**Comment by Simran-B** [comment]
<p>Not sure if it helps, but I want to add that if you manually invoke runs (<code>workflow_dispatch</code>), then you can define input fields and pass arguments as desired.</p>

**Comment by ghostwriter** [comment]
<p>You can set an environment variable using <code>set-env</code> at the beginning of the steps:</p>
<pre><code class="lang-auto">- run: echo ::set-env name=RELEASE_NAME::$(if [ "${{ secrets.RELEASE_NAME }}" ]; then echo ${{ secrets.RELASE_NAME }}; else echo "$GITHUB_REPOSITORY" | awk -F / '{print $2}'; fi)
</code></pre>
<p>Then use it in the next steps:</p>
<pre><code class="lang-auto">ghcr.io/${{ github.repository_owner }}/${{ env.RELEASE_NAME }}
</code></pre>
<p>Hope that helps. :v:t4:</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/eiymba">@eiymba</a> ,</p>
<p>In the workflow, you can use the <a href="https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables" rel="noopener nofollow ugc">default environment variable</a> ‘<strong>GITHUB_REPOSITORY</strong>’ to get the owner and repository name (<strong>owner-name/repo-name</strong>).<br>
If you want only get the repository name, you use the following method to extract it from ‘<strong>GITHUB_REPOSITORY</strong>’.</p>
<pre><code class="lang-yaml">- name: get repository name
  run: |
    echo $GITHUB_REPOSITORY
    repo_name=${GITHUB_REPOSITORY##*/}
    echo $repo_name
</code></pre>
<p>If you want to only get the owner name, you can use the expression ‘<strong>${{ github.repository_owner }}</strong>’ to get it from the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="noopener nofollow ugc">github context</a>.</p>
<pre><code class="lang-yaml">- name: get owner name
  run: |
    owner_name=${{ github.repository_owner }}
    echo $owner_name
</code></pre>
<p>The following is an example as reference: <a href="https://github.com/TestWorkflowsRML/GHA-Syntax/actions/runs/264370292/workflow#L21" rel="noopener nofollow ugc">https://github.com/TestWorkflowsRML/GHA-Syntax/actions/runs/264370292/workflow#L21</a></p>

**Comment by eiymba** [comment]
<p>Hi <a class="mention" href="/ghostwriter">@ghostwriter</a> ! Thanks a lot of taking a look at this. &lt;3</p>
<p>I’ve tried this solution before but the build process does not appear to echo secrets when used in this manner. It’s entirely omitted for me.</p>
<p>Give us a shout if you have any more ideas! Thanks again!</p>

**Comment by eiymba** [comment]
<p>Hey <a class="mention" href="/brightran">@brightran</a> thanks a lot for taking a look at this! &lt;3</p>
<p>This is currently the solution I’m already using. I’m looking to use an arbitrary name at build time instead of the repository name, without writing it in the yaml file, and setting this as a secret does not appear to work. For example, if my repository is called <code>my-app-docker</code>, I’d want to name my docker image <code>my-app</code> instead of <code>my-app-docker</code>.</p>
<p>Give us a shout if you have any more ideas! Thanks again!</p>

**Comment by eiymba** [comment]
<p>This might be the best approach thus far! It works but the only down side is, it’s not automatically triggered.</p>
<p>Thanks for taking a look at this! &lt;3</p>

**Comment by eiymba** [comment, marked as answer]
<p>I’ve found a solution.</p>
<p>I create a separate workflow which generates an artefact, and then I can check to see if it’s available for my release workflow. to consume.</p>
<p>I’ll make a write up when I have time.</p>
<p>Thanks for your help, everyone!</p>
